src/compiler/parser.ts

```diff
@@ -4538,6 +4538,10 @@ namespace ts {
                 // isn't actually allowed, but we want to treat it as a lambda so we can provide
                 // a good error message.
                 if (isModifierKind(second) && second !== SyntaxKind.AsyncKeyword && lookAhead(nextTokenIsIdentifier)) {

```

Comment:
The fact that `nextTokenIsIdentifier` checks for `async` and `yield` itself makes me feel like there should be some sort of context for `as`, rather than special casing, but like you said on https://github.com/microsoft/TypeScript/issues/44466#issuecomment-1121378110, I'm not entirely certain that there's another expression that can behave like this.

This whole "does this look like an arrow function" is effectively a heuristic and continues to bite us (see: #16241), so I think it's likely ok to disable this heuristic here like you've done.

Reply:
I think the other option is reworking this along the lines Daniel suggested earlier in #44466, but looking for a matched close paren followed by an arrow? But doing that leaves a lot of lookahead for errors to accumulate, and you're not guaranteed to find the matching paren, much less the arrow itself.

Reply:
Yeah, I think that this fix is good as-is; this block (as commented above) is to do better error recovery if someone writes code like `(foo number) => ...`, which is guaranteed to fail to parse later. The likelihood of that broken code appearing _and_ someone legitimately naming a type `as` seems very, very low, so I'm happy with just not having recovery in this case.

Reply:
I suppose though, that one could write a test that is `(foo as) => 1234` or something, and see if you should actually return `TriState.Unknown` here instead.

Or, that this should be a `lookAhead` in the if statement before the code you've added (the "isn't actually allowed" one), to better fit with the heuristic.

Reply:
I have tested both, and they both pass the tests in the PR, though they all differ on a test like:

```ts
let out: any = undefined;
type as = any;

export const e = (declare as) => 1234;
```

With the two lookahead versions, you get an emitted `exports.e = declare;`, but with `Unknown`, it actually tries to parse that broken code as an arrow function again and complains about the modifier, and emits:

```js
var e = function (as) { return 1234; };
exports.e = e;
```

Maybe that's better in the error case, preserving the recovery?

@DanielRosenwasser curious as to what you think; I'm not convinced it matters either way if both fix the bug, just an esoteric recovery.

Reply:
I think this is an okay fix; however, if we add `satisfies` (or any other new infix keyword) then we'll have another similar bug in the future.